### PR TITLE
INT-1962: Update Metric and Events script for CSP support and Readme

### PR DESCRIPTION
**Pull request by svinothini24**
[Validate Nagios Integration - after Python SDK updated with CSP Support](https://jira.eng.vmware.com/browse/INT-1962)

### CODE REVIEW

**Comment by svinothini24** [comment]
@gangadharaswamy Please review

**Comment by svinothini24** [comment]
@dkuldeep22 Please review

File: nagios/nagios-metrics-wf.py [hunk 1]
```diff
@@ -112,35 +113,74 @@ def remove_trailing_dots(text):
...
+            csp_org_id=args.csp_org_id,
+            max_queue_size=50000,
+            batch_size=10000,
+            flush_interval_seconds=5)
     else:
         parser.print_help()
         exit(-1)
 
+    wavefront_client = client_factory.get_client()
+    print(wavefront_client)
```
Comment (dkuldeep22) [inline nagios/nagios-metrics-wf.py:169]: is this print needed? or could use some logging.

File: nagios/nagios-events-wf.py [hunk 2]
```diff
@@ -1,55 +1,99 @@
...
+            csp_org_id=args.csp_org_id,
+            max_queue_size=50000,
+            batch_size=10000,
+            flush_interval_seconds=5)
+    else:
+        parser.print_help()
+        exit(-1)
+
+    wavefront_client = client_factory.get_client()
+    print(wavefront_client)
```
Comment (dkuldeep22) [inline nagios/nagios-events-wf.py:86]: same as below

File: nagios/nagios-events-wf.py [hunk 3]
```diff
@@ -1,55 +1,99 @@
...
+
+    try:
+        main()
+    finally:
+        # If the application failed to send metrics/histograms/tracing-spans,
+        # you can get the total failure count as follows:
+        total_failures = wavefront_client.get_failure_count()
```
Comment (dkuldeep22) [inline nagios/nagios-events-wf.py:93]: can we log failure counts?

File: nagios/README.md [hunk 4]
```diff
@@ -22,17 +22,37 @@ host_perfdata_file_processing_command=wavefront_perf_host
...
+                                                           --wf_server [WAVEFRONT_URL] \
+                                                           --csp_app_id [CSP_APP_ID] \
+                                                           --csp_app_secret [CSP_APP_SECRET] \
+                                                           --csp_base_url [CSP_BASE_URL]                                                           
 }
 
 define command {
     command_name  wavefront_perf_service
-    command_line  /opt/nagios/etc/wf/nagios-metrics-wf.py [PATH_PERFDATA]/service-perfdata \
+    command_line  /usr/local/nagios/libexec/nagios-metrics-wf.py [PATH_PERFDATA]/service-perfdata \
```
Comment (dkuldeep22) [inline nagios/README.md:41]: this will change if we are having multiple files with single command OR single command and rest as example?

Comment (svinothini24) [reply nagios/README.md:41]: Yes. I'll change the Readme also

**Review by gangadharaswamy — APPROVED** [review]

**Review by dkuldeep22 — APPROVED** [review]
LGTM

**Comment by svinothini24** [comment]
@mmihaylovam Please review

File: nagios/README.md [hunk 5]
```diff
@@ -9,36 +9,86 @@ service_perfdata_file_template=DATATYPE::SERVICEPERFDATA\tTIMET::$TIMET$\tHOSTNA
...
 host_perfdata_file=[PATH_PERFDATA]/host-perfdata
 host_perfdata_file_template=DATATYPE::HOSTPERFDATA\tTIMET::$TIMET$\tHOSTNAME::$HOSTNAME$\tHOSTPERFDATA::$HOSTPERFDATA$\tHOSTCHECKCOMMAND::$HOSTCHECKCOMMAND$\tHOSTSTATE::$HOSTSTATE$\tHOSTSTATETYPE::$HOSTSTATETYPE$
 
 host_perfdata_file_mode=a
 host_perfdata_file_processing_interval=15
-host_perfdata_file_processing_command=wavefront_perf_host
+host_perfdata_file_processing_command=wavefront-perf-host
 ```
 
+#### Using CSP API Token
```
Comment (mmihaylovam) [inline nagios/README.md:22]: #### Using a VMware Cloud Services API Token

File: nagios/README.md [hunk 6]
```diff
@@ -9,36 +9,86 @@ service_perfdata_file_template=DATATYPE::SERVICEPERFDATA\tTIMET::$TIMET$\tHOSTNA
...
+    command_name  wavefront-perf-service
+    command_line /usr/local/nagios/libexec/nagios-metrics-wf.py [PATH_PERFDATA]/service-perfdata \
+                                                           --service \
+                                                           --wf_server [WAVEFRONT_URL] \
+                                                           --csp_api_token [CSP_API_TOKEN] \
+                                                           --csp_base_url [CSP_BASE_URL]
+}
+```
+
+#### Using CSP Oauth
```
Comment (mmihaylovam) [inline nagios/README.md:42]: #### Using the Credentials of a Server to Server OAuth App in VMware Cloud Services

File: nagios/README.md [hunk 7]
```diff
@@ -9,36 +9,86 @@ service_perfdata_file_template=DATATYPE::SERVICEPERFDATA\tTIMET::$TIMET$\tHOSTNA
...
+    command_line /usr/local/nagios/libexec/nagios-metrics-wf.py [PATH_PERFDATA]/service-perfdata \
+                                                           --service \
+                                                           --wf_server [WAVEFRONT_URL] \
+                                                           --csp_app_id [CSP_APP_ID] \
+                                                           --csp_app_secret [CSP_APP_SECRET] \
+                                                           --csp_base_url [CSP_BASE_URL]
+}
+```
+
+#### Using Wavefront Ingestion
```
Comment (mmihaylovam) [inline nagios/README.md:64]: #### Using an Operations for Applications API Token

File: nagios/README.md [hunk 8]
```diff
@@ -9,36 +9,86 @@ service_perfdata_file_template=DATATYPE::SERVICEPERFDATA\tTIMET::$TIMET$\tHOSTNA
...
+    command_name  wavefront-perf-service
+    command_line  /usr/local/nagios/libexec/nagios-metrics-wf.py [PATH_PERFDATA]/service-perfdata \
+                                                           --service \
+                                                           --wf_server [WAVEFRONT_URL] \
+                                                           --wf_token [WAVEFRONT_TOKEN]
 }
 ```
 
-### Events
+Refer the configuration samples at https://github.com/wavefrontHQ/integrations/raw/master/nagios/config to send metrics using CSP API token, CSP Oauth or Wavefront API Token.
```
Comment (mmihaylovam) [inline nagios/README.md:83]: ...using a VMware Cloud services API token, server to server OAuth app credentials, or an Operations for Applications API token.

File: nagios/README.md [hunk 9]
```diff
@@ -9,36 +9,86 @@ service_perfdata_file_template=DATATYPE::SERVICEPERFDATA\tTIMET::$TIMET$\tHOSTNA
...
+
+### Contacts
+
+Add the below Contact and Contact group to the `contacts.cfg`.
+
 ```
-define contact{
+# Wavefront contact definition for event notification commands
```
Comment (mmihaylovam) [inline nagios/README.md:90]: Operations for Applications contact definition....

File: nagios/README.md [hunk 10]
```diff
@@ -51,39 +101,99 @@ define contact{
   host_notification_commands      nagios-to-wavefront-host
 }
 
-define contact{
+# Wavefront user definition
```
Comment (mmihaylovam) [inline nagios/README.md:104]: Operations for Applications user definition

File: nagios/README.md [hunk 11]
```diff
@@ -51,39 +101,99 @@ define contact{
...
   use             wf-generic-contact
-  alias           Sam
-  email           sam@foo.com
+  alias           Admin
+  email           admin@example.com
   address1        +155512312
 }
 
-define contactgroup{
+# Contact group for Wavefront user definition
```
Comment (mmihaylovam) [inline nagios/README.md:113]: ....Operations for Applications user definition

File: nagios/README.md [hunk 12]
```diff
@@ -51,39 +101,99 @@ define contact{
...
+define contactgroup {
   contactgroup_name  wavefront
```
Comment (mmihaylovam) [inline nagios/README.md:115]: Can we use the product name, for example: 
```
 contactgroup_name  operations-for-applications
  alias              Notifications sent to Operations for Applications
  members            operations-for-applications-user
```

File: nagios/README.md [hunk 13]
```diff
@@ -51,39 +101,99 @@ define contact{
...
   alias              Notifications send to wavefront
   members            wfuser
 }
+```
+
+### Events
+#### Using CSP API Token
```
Comment (mmihaylovam) [inline nagios/README.md:122]: Using a VMware Cloud Services API Token

File: nagios/README.md [hunk 14]
```diff
@@ -51,39 +101,99 @@ define contact{
...
 
-define command{
-  command_name nagios-to-wavefront-host
-  command_line /opt/nagios/etc/wf/nagios-events-wf.py --type '$NOTIFICATIONTYPE$' \
-                                                      --host '$HOSTNAME$' \
-                                                      --time '$TIMET$' \
-                                                      --msg '$HOSTOUTPUT$' \
-                                                      [WAVEFRONT_URL] \
-                                                      [WAVEFRONT_TOKEN]
+#### Using CSP Oauth
```
Comment (mmihaylovam) [inline nagios/README.md:147]: #### Using the Credentials of a Server to Server OAuth App in VMware Cloud Services

File: nagios/README.md [hunk 15]
```diff
@@ -51,39 +101,99 @@ define contact{
...
+                                                        --host '$HOSTNAME$' --time '$TIMET$' \
+                                                        --msg '$SERVICEOUTPUT$\n$NOTIFICATIONAUTHOR$\n$NOTIFICATIONCOMMENT$' \
+                                                        --wf_server [WAVEFRONT_URL] \
+                                                        --csp_app_id [CSP_APP_ID] \
+                                                        --csp_app_secret [CSP_APP_SECRET] \
+                                                        --csp_base_url [CSP_BASE_URL]
+}
+ ```
+
+#### Using Wavefront Ingestion
```
Comment (mmihaylovam) [inline nagios/README.md:174]: #### Using an Operations for Applications API Token

File: nagios/README.md [hunk 16]
```diff
@@ -51,39 +101,99 @@ define contact{
...
+    command_line /usr/local/nagios/libexec/nagios-events-wf.py -S --type '$NOTIFICATIONTYPE$' \
+                                                       --service '$SERVICEDISPLAYNAME$' \
+                                                       --host '$HOSTNAME$' --time '$TIMET$' \
+                                                       --msg '$SERVICEOUTPUT$\n$NOTIFICATIONAUTHOR$\n$NOTIFICATIONCOMMENT$' \
+                                                       --wf_server [WAVEFRONT_URL] \
+                                                       --wf_token [WF_API_TOKEN]
+}
+```
+
+Refer the configuration samples at https://github.com/wavefrontHQ/integrations/raw/master/nagios/config to send events using CSP API token, CSP Oauth or Wavefront API Token.
```
Comment (mmihaylovam) [inline nagios/README.md:198]: ...using a VMware Cloud services API token, server to server OAuth app credentials, or an Operations for Applications API token.

File: nagios/config/wavefront-csp-oauth.cfg [hunk 17]
```diff
@@ -0,0 +1,45 @@
+# Nagios command definitions for host and service metrics to the Wavefront server using CSP Oauth
```
Comment (mmihaylovam) [inline nagios/config/wavefront-csp-oauth.cfg:1]: Nagios command definitions for host and service metrics to the Operations for Applications server using server to server OAuth app credentials (id and secret)

File: nagios/config/wavefront-csp-oauth.cfg [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    command_name  wavefront-perf-service
+    command_line /usr/local/nagios/libexec/nagios-metrics-wf.py /usr/local/nagios/var/service-perfdata \
+                                                        --service \
+                                                        --wf_server [WAVEFRONT_URL] \
+                                                        --csp_app_id [CSP_APP_ID] \
+                                                        --csp_app_secret [CSP_APP_SECRET] \
+                                                        --csp_base_url [CSP_BASE_URL]
+}
+
+# Wavefront contact definition for event notification commands using CSP Oauth
```
Comment (mmihaylovam) [inline nagios/config/wavefront-csp-oauth.cfg:22]: Operations for Applications contact definition for event notification commands using the credentials (id and secret) of a server to server OAuth app in VMware Cloud services.

File: nagios/config/wavefront-csp-token.cfg [hunk 19]
```diff
@@ -0,0 +1,41 @@
+# Nagios command definitions for host and service metrics to the Wavefront server using CSP API Token
```
Comment (mmihaylovam) [inline nagios/config/wavefront-csp-token.cfg:1]: Nagios command definitions for host and service metrics to the Operations for Applications server using a VMware Cloud services API token

File: nagios/config/wavefront-csp-token.cfg [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+define command {
+    command_name  wavefront-perf-service
+    command_line  /usr/local/nagios/libexec/nagios-metrics-wf.py /usr/local/nagios/var/service-perfdata \
+                                                        --service \
+                                                        --wf_server [WAVEFRONT_URL] \
+                                                        --csp_api_token [CSP_API_TOKEN] \
+                                                        --csp_base_url [CSP_BASE_URL]
+}
+
+# Wavefront contact definition for event notification commands using CSP API Token
```
Comment (mmihaylovam) [inline nagios/config/wavefront-csp-token.cfg:20]: Operations for Applications contact definition for event notification commands using a VMware Cloud services API token

File: nagios/config/wavefront-ingestion.cfg [hunk 21]
```diff
@@ -0,0 +1,39 @@
+# Nagios command definitions for host and service metrics to the Wavefront server
```
Comment (mmihaylovam) [inline nagios/config/wavefront-ingestion.cfg:1]: Nagios command definitions for host and service metrics to the Operations for Applications server using an  Operations for Applications API token

File: nagios/config/wavefront-ingestion.cfg [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+
+define command {
+    command_name  wavefront-perf-service
+    command_line  /usr/local/nagios/libexec/nagios-metrics-wf.py /usr/local/nagios/var/service-perfdata \
+                                                       --service \
+                                                       --wf_server [WAVEFRONT_URL] \
+                                                       --wf_token [WAVEFRONT_TOKEN]
+}
+
+# Wavefront contact definition for event notification commands using Wavefront API Token
```
Comment (mmihaylovam) [inline nagios/config/wavefront-ingestion.cfg:18]: Operations for Applications contact definition for event notification commands using an Operations for Applications API token

File: nagios/nagios-events-wf.py [hunk 23]
```diff
@@ -1,55 +1,100 @@
...
-        print("OK")
-        sys.exit(0)
+    wavefront_client.send_event(name, startTime, endTime, args.host, ["env:", "test-event"], annotations)
 
 
 if __name__ == "__main__":
-    parser = argparse.ArgumentParser()
-    parser.add_argument('server', type=str, nargs=1, help='Wavefront Server URL.')
-    parser.add_argument('token', type=str, nargs=1, help='Wavefront token.')
+    parser = argparse.ArgumentParser(description='Sends Nagios perfdata to wavefront.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:26]: Sends Nagios perfdata to Operations for Applications.

File: nagios/nagios-events-wf.py [hunk 24]
```diff
@@ -1,55 +1,100 @@
...
+    parser.add_argument('--wf_server', dest="server", help='Wavefront Server URL.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:27]: help='Operations for Applications server URL.'

File: nagios/nagios-events-wf.py [hunk 25]
```diff
@@ -1,55 +1,100 @@
...
+    parser.add_argument('--wf_token', dest="token", help='Wavefront token.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:28]: help='Operations for Applications API token.')

File: nagios/nagios-events-wf.py [hunk 26]
```diff
@@ -1,55 +1,100 @@
...
+    parser.add_argument('--type', help='Notification type.')
+    parser.add_argument('--host', help='Notification hostname.')
     parser.add_argument('--service', type=str, nargs='?', help='Notification service name.')
-    parser.add_argument('--time', type=int, nargs=1, help='Notification time (UNIX epoch).')
-    parser.add_argument('--msg', type=str, nargs=1, help='Notification message.')
+    parser.add_argument('--time', type=int, help='Notification time (UNIX epoch).')
+    parser.add_argument('--msg', help='Notification message.')
+    parser.add_argument('--wf_proxy_addr', dest="addr")
+    parser.add_argument('--wf_proxy_port', dest="port")
+    parser.add_argument('--csp_base_url', dest="csp_base_url", help='Base URL of CSP console.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:37]: help='Base URL of the VMware Cloud Services Console.'

Isn't this static? Can we provide the actual URL?

File: nagios/nagios-events-wf.py [hunk 27]
```diff
@@ -1,55 +1,100 @@
...
+    parser.add_argument('--csp_api_token', dest="csp_api_token", help='CSP API Token for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:38]: help='VMware Cloud services API token.'

File: nagios/nagios-events-wf.py [hunk 28]
```diff
@@ -1,55 +1,100 @@
...
+    parser.add_argument('--csp_app_id', dest="csp_app_id", help='CSP OAuth App ID for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:39]: ' ID of a server to server OAuth app in VMware Cloud services.

File: nagios/nagios-events-wf.py [hunk 29]
```diff
@@ -1,55 +1,100 @@
...
+    parser.add_argument('--csp_app_secret', dest="csp_app_secret", help='CSP OAuth App secret for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:40]: ' Secret of a server to server OAuth app in VMware Cloud services.

File: nagios/nagios-events-wf.py [hunk 30]
```diff
@@ -1,55 +1,100 @@
...
+    parser.add_argument('--csp_org_id', dest="csp_org_id", help='CSP Organization ID for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-events-wf.py:41]: help='Long ID of the VMware Cloud organization running the service. The OAuth app must be added to this organization.'

File: nagios/nagios-metrics-wf.py [hunk 31]
```diff
@@ -112,35 +113,75 @@ def remove_trailing_dots(text):
     parser.add_argument('--service', dest='service', action='store_true',
                         help='proccess "service-perfdata" (default: "host-perfdata")')
     parser.add_argument('--test', dest='test', action='store_true')
-    parser.add_argument('--wf_server', dest="server")
-    parser.add_argument('--wf_token', dest="token")
-    parser.add_argument('--wf_proxy_addr', dest="addr")
-    parser.add_argument('--wf_proxy_port', dest="port")
+    parser.add_argument('--wf_server', dest="server", help='Wavefront Server URL.')
```
Comment (mmihaylovam) [inline nagios/nagios-metrics-wf.py:116]: Operations for Applications server URL.

File: nagios/nagios-metrics-wf.py [hunk 32]
```diff
@@ -112,35 +113,75 @@ def remove_trailing_dots(text):
...
+    parser.add_argument('--wf_token', dest="token", help='Wavefront token.')
```
Comment (mmihaylovam) [inline nagios/nagios-metrics-wf.py:117]: help='Operations for Applications API token.'

File: nagios/nagios-metrics-wf.py [hunk 33]
```diff
@@ -112,35 +113,75 @@ def remove_trailing_dots(text):
...
+    parser.add_argument('--wf_proxy_addr', dest="addr", help='Wavefront Proxy address.')
+    parser.add_argument('--wf_proxy_port', dest="port", help='Wavefront Proxy port.')
+    parser.add_argument('--csp_base_url', dest= "csp_base_url", help='Base URL of CSP console.')
```
Comment (mmihaylovam) [inline nagios/nagios-metrics-wf.py:120]: help='Base URL of the VMware Cloud Services Console.'

Can we provide the actual URL if it's static?

File: nagios/nagios-metrics-wf.py [hunk 34]
```diff
@@ -112,35 +113,75 @@ def remove_trailing_dots(text):
...
+    parser.add_argument('--csp_api_token', dest="csp_api_token", help='CSP API Token for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-metrics-wf.py:121]: help='VMware Cloud services API token.'

File: nagios/nagios-metrics-wf.py [hunk 35]
```diff
@@ -112,35 +113,75 @@ def remove_trailing_dots(text):
...
+    parser.add_argument('--csp_app_id', dest="csp_app_id", help='CSP OAuth App ID for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-metrics-wf.py:122]: help='ID of a server to server OAuth app in VMware Cloud services.

File: nagios/nagios-metrics-wf.py [hunk 36]
```diff
@@ -112,35 +113,75 @@ def remove_trailing_dots(text):
...
+    parser.add_argument('--csp_app_secret', dest="csp_app_secret", help='CSP OAuth App secret for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-metrics-wf.py:123]: help='Secret of a server to server OAuth app in VMware Cloud services.

File: nagios/nagios-metrics-wf.py [hunk 37]
```diff
@@ -112,35 +113,75 @@ def remove_trailing_dots(text):
...
+    parser.add_argument('--csp_org_id', dest="csp_org_id", help='CSP Organization ID for csp authentication.')
```
Comment (mmihaylovam) [inline nagios/nagios-metrics-wf.py:124]: help='Long ID of the VMware Cloud organization running the service. The OAuth app must be added to this organization.'

**Review by mmihaylovam — CHANGES_REQUESTED** [review]
Please take a look at my comments. We don't use Wavefront and CSP.

File: nagios/nagios-events-wf.py (hunk 26 above)
Comment (svinothini24) [reply nagios/nagios-events-wf.py]: As the URL varies for staging and production, I did not hardcode the actual URL.
@gangadharaswamy Please share your thoughts

File: nagios/nagios-metrics-wf.py (hunk 33 above)
Comment (svinothini24) [reply nagios/nagios-metrics-wf.py]: As the URL varies for staging and production, I did not hardcode the actual URL.
@gangadharaswamy Please share your thoughts

Comment (gangadharaswamy) [reply nagios/nagios-metrics-wf.py]: @svinothini24 
Let's use `https://console.cloud.vmware.com` for the CSP Base URL. This is for production clusters.
For internal use, we can modify the url.

Comment (svinothini24) [reply nagios/nagios-metrics-wf.py]: can we remove this parameter `--csp_base_url` itself for the user?

Comment (svinothini24) [reply nagios/nagios-metrics-wf.py]: In order to be flexible with the csp_base_url, as that impacts the changes on cfg also, we will get the base url from the command line itself than static

**Review by mmihaylovam — APPROVED** [review]
LGTM

**Pull request merged**